SU2_CFD/include/numerics/CNumerics.hpp

```diff
@@ -659,6 +627,58 @@ class CNumerics {
     }
   }
 
+  /*!
+   * \brief Project average gradient onto normal (with or w/o correction) for viscous fluxes of scalar quantities.
+   * \param[in] nDim - Dimension of the flow problem, 2 or 3.
+   * \param[in] nVar - Number of variables.
+   * \param[in] normal - Area vector.
+   * \param[in] coord_i - Coordinate of point i.
+   * \param[in] coord_j - Coordinate of point j.
+   * \param[in] grad_i - Gradient at point i.
+   * \param[in] grad_j - Gradient at point j.
+   * \param[in] correct - Correct
+   * \param[in] var_i - Variable at point i.
+   * \param[in] var_j - Variable at point j.
+   * \param[out] projNormal - Average gradient projected on normal.
+   * \param[out] projCorrected - Corrected gradient projected on normal.
+   * \return (Edge_Vector DOT normal) / |Edge_Vector|^2.
+   */
+  template<class Vec1, class Vec2, class Mat>
+  FORCEINLINE static su2double ComputeProjectedGradient(int nDim, int nVar, const Vec1& normal,
+                                                        const Vec1& coord_i, const Vec1& coord_j,
+                                                        const Mat& grad_i, const Mat& grad_j,
+                                                        bool correct,
+                                                        const Vec2& var_i, const Vec2& var_j,
+                                                        su2double* projNormal,
+                                                        su2double* projCorrected) {
+    nDim = (nDim > 2)? 3 : 2;
```

Comment:
This keeps the compiler from going crazy with loop unrolling thinking that nDim is large, and then creating unrolled code that is never used.

Reply:
I should add an assert though, for 2 or 3.
Yes compiler explorer

Reply:
Of course this is a valid check, but can you motivate that a bit why you think it is additionally necessary here? I guess if we would do this kind of sanity checks of sizes in more places, we would have quite a bit of them. At first glance that looks like a bit overkill but maybe i just dont understand it.

And might this already be enough for the compiler to not unroll further because it puts a hard constraint on the nDim value

Reply:
Because the code is only right if called with nDim = 2 or 3, and it is good practice to assert that kind of assumption to facilitate debugging, keep in mind that asserts are no-ops in release builds!
This is also why they cannot be used to inform the compiler that we expect 2 or 3 -> https://gcc.godbolt.org/z/nbPv8KhPo